### DISCUSSION TITLE
Background images?

### DISCUSSION THREAD

**Discussion by janert** [bbidulock/icewm#665] [Q&A] [answered]
I was excited to read in the man page that iceWM now supports a different background for each workspace. That is a feature I have been pining for for a long time!

But I could not find any instructions how to set this up. Is there a "switchWorkspaceHook" option or something like that? 

In fact, right now I am having difficulties letting iceWM handle my background image at all. I can set it, upon startup, in my preferences file. But when I try to change it using "icewmbg -i ...", I end up with a solid color background (which is I believe the default for my current theme). What am I doing wrong; how do I debug that?

icewm -V
IceWM 2.9.6, Copyright 1997-2012 Marko Macek, 2001 Mathias Hasselmann.

icewm --configured
icewm configured options: ao fribidi i18n imlib2 librsvg nls session shape xfreetype2 xinerama xrandr

Operating System: Linux Mint 21                   
Kernel: Linux 5.15.0-46-generic

**Comment by gijsbers** [comment]
You need to differentiate the command line from the configuration file.

1. `icewmbg -i "/path/to/wallpaper*,/more/image*" -p -u`
   Note the quotes around the two image patterns.
2. In `~/.icewm/prefoverride` say:
   ```
   DesktopBackgroundImage=/path/to/wallpaper*,/more/image*
   ShuffleBackgroundImages=1
   ```
   Then `icewmbg -r`.

It is all explained in the manpage https://ice-wm.org/man/icewmbg.

**Comment by janert** [comment]
Thanks for the fast reply! 

What I am taking away here is that I can't assign a specific image to a specific workspace, but that they get "shuffled" randomly when I change workspaces. I that correct?

I will admit that this is not what I had hoped for. My idea was to assign a specific image to a specific workspace.

And I don't believe this is explained in the manpage. Where does it say that icewmbg runs when I change workspaces? (I am only aware of the timeout version.) Maybe it's there, and I am missing it.

And, running the danger of seeming dense: I would assume that running icewmbg from the command line would take precedence over the preferences file and "just set the background". In my case it doesn't. When I run it (as you say), I get neither the bg named in the command line, nor the one in my preferences file, but the theme's default. Creating a prefoverride file does not make a difference.

**Comment by gijsbers** [comment]
Shuffling is just an option.

You are correct that per-workspace could be clarified in the manpage.

If I do `icewmbg -i /some/image -p`, then it overrides the theme default.

If I set `DesktopBackgroundImage=/some/image` in prefoverride and do `icewmbg -r` then it shows /some/image.

If you want to control which image for each workspace, just say
`icewmbg -i image0,image1,image2,image3 -p`

Or `DesktopBackgroundImage=image0,image1,image2,image3` in prefoverride and restart icewmbg.

**Comment by janert** [comment, marked as answer]
I managed to figure it out... (finally).

1) Thanks for explaining the comma-separated-list logic. It had not occurred to me that icewm simply assigns the images to workspaces in the order in which they appear - although it makes total sense, once you realize this.

2) When using icewmbg -i path-to-img, I never got the image, only the theme's default bg color... until I finally realized that the path must be ABSOLUTE! Once I realized this, it works fine. (Moreover, if there are multiple images, icewmbg does not terminate, so one should put it in the background. This was not obvious to me; took me a while to catch up to that.) 

It might be nice if icewmbg printed out some form of error msg if it can't find the image. (Admittedly, not clear where the msg should go if it is run automatically.) 

The man page might also be more explicit about the need for absolute paths; and the fact that icewmbg needs to be kept running as a daemon to manage the bg when switching workspaces.)

Anyway, thanks again for your help. I am glad to (finally!) have figured this out.
